File: frontend/src/types/experiment.ts
export interface ExperimentBatch {
  id: number
  name: string
  description: string
  remark?: string
  selected_dates: string[]
  selected_patient_ids: string[]
  status: string
  total_tasks: number
  success_count: number
  failure_count: number
  patient_count?: number
  combination_count?: number
  created_at: string | Date
  updated_at: string | Date
  started_at?: string | Date | null
  completed_at?: string | Date | null
  metrics?: ExperimentMetrics
  combinations?: ExperimentCombination[]
  asr_models?: string[]
  llm_models?: string[]
  prompt_templates?: string[]
  asr_reuse_count?: number
  asr_generated_count?: number
  asr_failed_count?: number
  asr_reuse_rate?: number
}

export interface ExperimentCombination {
  id: number
  batch_id: number
  asr_model_id: number
  llm_model_id: number | null
  prompt_name: string
  prompt_template: string
  hotwords: string[]
  enabled: boolean
  created_at: string | Date
  asr_model_name?: string
  llm_model_name?: string
}

export interface ExperimentTaskSummary {
  id: number
  batch_id: number
  combination_id: number
  patient_id: number
  record_id?: string
  date?: string
  exam_record_id?: number
  stage: string
  status: string
  retry_count: number
  accuracy: number | null
  total_duration: number
  error_type: string | null
  error_message?: string | null
  created_at: string | Date

  // 快照字段
  asr_result_id?: number | null
  llm_result_id?: number | null
  asr_source?: string | null
  asr_model_name?: string | null
  llm_model_name?: string | null
  prompt_template_name?: string | null
  full_transcript?: string
  asr_results?: any[]
  llm_raw_output?: string | null
  structured_result?: any
  summary_text?: string | null
  evaluation?: any
  ground_truth?: any

  // 组合信息
  combination_asr_name?: string
  combination_llm_name?: string
  combination_prompt_name?: string
}

export interface ExperimentMetrics {
  total_tasks: number
  success_count: number
  failure_count: number
  patient_count?: number
  asr_success_rate: number
  asr_empty_rate: number
  avg_asr_duration: number
  field_accuracy: {
    endometrium_thickness: number
    endometrium_type: number
    ovary_size: number
    follicle: number
    remark: number
  }
  complete_patient_rate: number
  llm_failure_rate: number
  total_cost: number
}

File: backend/app/schemas/experiment.py
"""
实验批量编排 API Schema
"""
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime


class ExperimentBatchCreate(BaseModel):
    name: str
    description: str = ""
    remark: str = ""
    selected_dates: list[str] = Field(default_factory=list)
    selected_patient_ids: list[str] = Field(default_factory=list)
    # 单一组合配置（新设计：一个实验只允许一个组合）
    asr_model_id: int | None = None
    llm_model_id: int | None = None
    prompt_template_id: int | None = None
    prompt_name: str = ""
    prompt_template: str = ""
    hotwords: list[str] = Field(default_factory=list)


class ExperimentPatientScopeUpdate(BaseModel):
    selected_patient_ids: list[str] = Field(default_factory=list)


class ExperimentControlAction(BaseModel):
    pass  # No body needed for start/pause/resume/cancel/retry


class ExperimentBatchOut(BaseModel):
    id: int
    name: str
    description: str
    remark: str = ""
    selected_dates: list[str] = Field(default_factory=list)
    selected_patient_ids: list[str] = Field(default_factory=list)
    status: str
    total_tasks: int = 0
    success_count: int = 0
    failure_count: int = 0
    created_at: datetime
    updated_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ExperimentCombinationCreate(BaseModel):
    asr_model_id: int
    llm_model_id: int | None = None
    prompt_name: str = ""
    prompt_template: str = ""
    hotwords: list[str] = Field(default_factory=list)


class ExperimentCombinationUpdate(BaseModel):
    asr_model_id: int | None = None
    llm_model_id: int | None = None
    prompt_name: str | None = None
    prompt_template: str | None = None
    hotwords: list[str] | None = None
    enabled: bool | None = None


class ExperimentCombinationOut(BaseModel):
    id: int
    batch_id: int
    asr_model_id: int
    llm_model_id: int | None
    prompt_name: str
    prompt_template: str
    hotwords: list[str]
    enabled: bool
    created_at: datetime

    class Config:
        from_attributes = True


class ExperimentTaskSummary(BaseModel):
    id: int
    batch_id: int
    combination_id: int
    patient_id: int
    stage: str
    status: str
    retry_count: int
    accuracy: float | None
    total_duration: float
    error_type: str | None
    created_at: datetime

    class Config:
        from_attributes = True


class ExperimentMetrics(BaseModel):
    total_tasks: int
    success_count: int
    failure_count: int
    asr_success_rate: float
    asr_empty_rate: float
    avg_asr_duration: float
    right_follicle_accuracy: float
    left_follicle_accuracy: float
    endometrium_thickness_accuracy: float
    endometrium_type_accuracy: float
    right_ovary_accuracy: float
    left_ovary_accuracy: float
    complete_patient_rate: float
    llm_failure_rate: float
    total_cost: float


class ExperimentDetailOut(ExperimentBatchOut):
    combinations: list[ExperimentCombinationOut] = Field(default_factory=list)
    # 单一组合便捷字段
    asr_model_id: int | None = None
    llm_model_id: int | None = None
    prompt_name: str = ""
    prompt_template_name: str = ""
    hotwords: list[str] = Field(default_factory=list)


class ExperimentListResponse(BaseModel):
    id: int
    name: str
    status: str
    remark: str = ""
    selected_dates: list[str] = Field(default_factory=list)
    selected_patient_ids: list[str] = Field(default_factory=list)
    total_tasks: int
    success_count: int
    failure_count: int
    created_at: datetime
    # 单一组合展示字段
    patient_count: int = 0
    field_accuracy: dict = Field(default_factory=dict)
    asr_model_name: str = ""
    llm_model_name: str = ""
    prompt_template_name: str = ""

    model_config = {"protected_namespaces": (), "from_attributes": True}
